pydantic/type_adapter.py

```diff
@@ -233,100 +156,174 @@ def __init__(
+            for various ways to construct this namespace.
+
+            This works for the case where this function is called in a module that
+            has the target of forward references in its scope, but
+            does not always work for more complex cases.
+
+            For example, take the following:
+
+            a.py
+            ```python
+            from typing import Dict, List
+
+            IntList = List[int]
+            OuterDict = Dict[str, 'IntList']
+            ```
+
+            b.py
+            ```python test="skip"
+            from a import OuterDict
+
+            from pydantic import TypeAdapter
+
+            IntList = int  # replaces the symbol the forward reference is looking for
+            v = TypeAdapter(OuterDict)
+            v({'x': 1})  # should fail but doesn't
+            ```
+
+            If `OuterDict` were a `BaseModel`, this would work because it would resolve
+            the forward reference within the `a.py` namespace.
+            But `TypeAdapter(OuterDict)` can't determine what module `OuterDict` came from.
+
+            In other words, the assumption that _all_ forward references exist in the
+            module we are being called from is not technically always true.
+            Although most of the time it is and it works fine for recursive models and such,
+            `BaseModel`'s behavior isn't perfect either and _can_ break in similar ways,
+            so there is no right or wrong between the two.
+
+            But at the very least this behavior is _subtly_ different from `BaseModel`'s.
+        """
+        if not force and self._defer_build:
+            _mock_val_ser.set_type_adapter_mocks(self, str(self._type))
+            self._pydantic_complete = False
+            return False
+
         try:
-            self._core_schema = _getattr_no_parents(self._type, '__pydantic_core_schema__')
-            self._validator = _getattr_no_parents(self._type, '__pydantic_validator__')
-            self._serializer = _getattr_no_parents(self._type, '__pydantic_serializer__')
+            self.core_schema = _getattr_no_parents(self._type, '__pydantic_core_schema__')
+            self.validator = _getattr_no_parents(self._type, '__pydantic_validator__')
+            self.serializer = _getattr_no_parents(self._type, '__pydantic_serializer__')
         except AttributeError:
             config_wrapper = _config.ConfigWrapper(self._config)
             core_config = config_wrapper.core_config(None)
 
-            self._core_schema = _get_schema(self._type, config_wrapper, parent_depth=self._parent_depth)
-            self._validator = create_schema_validator(
-                schema=self._core_schema,
+            schema_generator = _generate_schema.GenerateSchema(config_wrapper, ns_resolver=ns_resolver)
+            intermediate_core_schema = schema_generator.generate_schema(self._type)
+            self.core_schema = schema_generator.clean_schema(intermediate_core_schema)
```

Comment:
Do we want to set mocks if we fail to generate the schema, similar to models?

Reply:
Yes, this is a good idea.